src/coreclr/jit/rangecheckcloning.cpp

```diff
@@ -0,0 +1,520 @@
+// This is similar to what the "Loop Cloning" phase does for loops. Example:
+//
+//   arr[i + 1] = x;
+//   arr[i + 3] = y;
+//   arr[i + 5] = z;
+//   arr[i + 8] = w;
+//
+// assertprop/rangecheck phases give up on the above bounds checks because of the
+// increasing offsets and there are no assertions that they can rely on.
+// This phase handles such cases by cloning the entire block (only the affected statements
+// to be precise) into "fast" and "slow" under a "cloned" condition:
+//
+//   if (i >= 0 && i < arr.Length - 8)
+//   {
+//       // Fast path
+//       arr[i + 1] = x; // no bounds check
+//       arr[i + 3] = y; // no bounds check
+//       arr[i + 5] = z; // no bounds check
+//       arr[i + 8] = w; // no bounds check
+//   }
+//   else
+//   {
+//       // Slow path
+//       arr[i + 1] = x; // bounds check
+//       arr[i + 3] = y; // bounds check
+//       arr[i + 5] = w; // bounds check
+//       arr[i + 8] = w; // bounds check
+//   }
+//
+// The phase scans all statements in a block and groups the bounds checks based on
+// "Base Index and Length" pairs (VNs). Then the phase takes the largest group and
+// clones the block to have fast and slow paths.
+
+//------------------------------------------------------------------------------------
+// Initialize: Initialize the BoundsCheckInfo with the given bounds check node.
+//    and perform some basic legality checks.
+//
+// Arguments:
+//    comp             - The compiler instance
+//    statement        - The statement containing the bounds check
+//    bndChkNode       - The bounds check node
+//    bndChkParentNode - The parent node of the bounds check node (either null or COMMA)
+//
+// Return Value:
+//    true if the initialization was successful, false otherwise.
+//
+bool BoundsCheckInfo::Initialize(const Compiler*   comp,
+                                 Statement*        statement,
+                                 GenTreeBoundsChk* bndChkNode,
+                                 GenTree*          bndChkParentNode)
+{
+    idxVN = comp->vnStore->VNConservativeNormalValue(bndChkNode->GetIndex()->gtVNPair);
+    lenVN = comp->vnStore->VNConservativeNormalValue(bndChkNode->GetArrayLength()->gtVNPair);
+    if ((idxVN == ValueNumStore::NoVN) || (lenVN == ValueNumStore::NoVN))
+    {
+        return false;
+    }
+
+    stmt         = statement;
+    bndChk       = bndChkNode;
+    bndChkParent = bndChkParentNode;
+
+    if (bndChkParent != nullptr)
+    {
+        assert(bndChkParent->OperIs(GT_COMMA));
+        if (bndChkParent->gtGetOp2() == bndChkNode)
+        {
+            // GT_BOUNDS_CHECK is mostly LHS of COMMAs.
+            // In rare cases, it can be either a root node or RHS of COMMAs
+            // Unfortunately, optRemoveRangeCheck doesn't know how to handle it
+            // being RHS of COMMAs. TODO-RangeCheckCloning: fix that
+            return false;
+        }
+    }
```

Comment:
Is this phase using range check still? Seems like this can be removed?

Reply:
hm.. not sure I understand the question. This phase is ran in the main opt loop, bounds checks are removed in codegen.
If you asking whether we can see such an IR then the answer is yes (a few hits accross all collections) - I just didn't want to touch the existing optRemoveRangeCheck as part of this PR, could be done separately (to inspect the diffs)